feat: add Weather Insurance extension guide (#1432)
flare-foundation/developer-hub: docs/fcc/guides/02-weather-insurance.mdx

---
title: Weather Insurance Extension
slug: weather-insurance-extension

> authors: [nikerzetic]

Reviewer: ```suggestion
authors: [fassko]
```

description: Build and deploy an extension that settles rainfall insurance.
tags: [intermediate, tee, ethereum]
keywords:
  [tee, trusted-execution-environment, extension, flare-network, coston2, fcc]
sidebar_position: 2
unlisted: false
---

import CodeBlock from "@theme/CodeBlock";
import WeatherInsurance from "!!raw-loader!/examples/developer-hub-fcc-weather-insurance/contracts/WeatherInsurance.sol";

Build and deploy a **Trusted Execution Environment (TEE) extension** that settles **parametric rainfall insurance** using weather data from [OpenWeatherMap](https://openweathermap.org/api) inside a secure enclave.
This guide walks through every step — from writing the smart contract and extension handler to deploying on Coston2 and running an end-to-end policy buy and settlement test.
The code for this example is available on [GitHub](https://github.com/flare-foundation/fce-weather-insurance).

:::info[New to Flare TEE?]

> A TEE extension is an offchain program that runs inside a Trusted Execution Environment.

Reviewer: ```suggestion
A TEE extension is an off-chain program that runs inside a Trusted Execution Environment.
```

    uint256 payout;
    uint256 premium;
    uint256 measuredMmE2;       // set on settle
    bool settled;
    bool paidOut;
    bool isPrivate;             // threshold held in TEE until settlement
    bytes32 termsCommitment;
    string lat;
    string lon;
    uint64 settlementUnlockAt;  // 00:00 UTC day after coverage date
}
```

Settlement unlock time is computed by the `SettlementTime` library — the earliest `requestSettlement` / `settle` call is midnight on the day after the coverage date.

### Buying a Policy

**Public buy** — all terms are visible on-chain:

```solidity title="contracts/WeatherInsurance.sol"
function buyPolicy(
    string calldata _date,
    uint256 _rainThresholdMmE2,
    uint256 _payout,
    uint256 _premium,
    string calldata _lat,
    string calldata _lon
) external returns (uint256 policyId)
```

The caller must `approve` the contract for `_premium` in `payToken` before calling.
The contract pulls the premium, reserves `_payout` from pool liquidity, and records the policy.

**Private buy** — coverage terms stay encrypted on-chain:

1. The buyer ECIES-encrypts ABI-encoded `PrivateBuyParams` with the Flare Confidential Compute extension public key.

> 2. `buyPolicyPrivate` sends the ciphertext as a `WEATHER` / `BUY` instruction.

Reviewer: ```suggestion
2. The `buyPolicyPrivate` function sends the ciphertext as a `WEATHER` / `BUY` instruction.
```

    }));
    // sendInstructions with opCommand = SETTLE
}
```

Anyone can then call `settle()` with the TEE-signed `ActionResult` that is returned by the extension.
The contract reconstructs `ActionResult.Hash()`, recovers the signer, and requires that the signer equal the registered `teeAddress`.
If measured precipitation meets the threshold, `payToken` is transferred to the policyholder.

### TEE Signature Verification

Both `relayPrivateBuy` and `settle` verify TEE results using the same pattern:

```solidity
bytes32 resultHash = keccak256(
    abi.encodePacked(
        keccak256(_resultData),
        _actionId,
        keccak256(bytes(_submissionTag)),
        _status
    )
);
address signer = _recover(_ethSigned(resultHash), _signature);
require(signer == teeAddress, "bad TEE signature");
```

The TEE node signs `ActionResult.Hash()` with the [EIP-191 personal-sign](https://eips.ethereum.org/EIPS/eip-191) prefix.
Only successful results are accepted.

### Extension ID Discovery

After the extension is registered on-chain, call `setExtensionId()` once to discover and cache the extension ID — same pattern as the [Private Key Extension](/fcc/guides/sign-extension).

> ## Offchain Handler

Reviewer: ```suggestion
## Off-chain Handler
```

> The offchain handler lives in `internal/extension/`.

Reviewer: ```suggestion
The off-chain handler lives in `internal/extension/`.
```

It exposes an HTTP server with two endpoints:

- **`GET /state`** — reports whether the OpenWeatherMap API key is configured.
- **`POST /action`** — receives TEE actions and routes them by `opType` / `opCommand`.

Constants match the Solidity contract:

```go title="internal/config/config.go"
const (
    OPTypeWeather   = "WEATHER"
    OPCommandFetch  = "FETCH"
    OPCommandSettle = "SETTLE"
    OPCommandBuy    = "BUY"
)
```

### FETCH Handler

> `processWeatherFetch` ABI-decodes a city name, calls the OpenWeatherMap API, and returns a JSON `WeatherReport` in `ActionResult.Data`:

Reviewer: ```suggestion
The `processWeatherFetch` function ABI-decodes a city name, calls the OpenWeatherMap API, and returns a JSON `WeatherReport` in `ActionResult.Data`:
```

```go title="internal/extension/extension.go"
func (e *Extension) processWeatherFetch(action teetypes.Action, df *instruction.DataFixed) teetypes.ActionResult {
    req, err := structs.Decode[types.GetWeatherRequest](types.GetWeatherMessageArg, df.OriginalMessage)
    // ...
    report, err := fetchWeather(req.City)
    payload, err := json.Marshal(report)
    return buildResult(action, df, payload, 1, nil)
}
```

### SETTLE Handler

> `processWeatherSettle` fetches daily precipitation from the OpenWeatherMap API and ABI-encodes the settlement:

Reviewer: ```suggestion
The `processWeatherSettle` function fetches daily precipitation from the OpenWeatherMap API and ABI-encodes the settlement:
```

```go title="internal/extension/extension.go"
encoded, err := types.SettlementResultArgs.Pack(
    req.ContractAddr,
    req.PolicyId,
    precipMmE2,
    coverageDate,
    revealedThreshold,
    triggered,
)
return buildResult(action, df, encoded, 1, nil)
```

For private policies, the handler loads coverage terms from in-enclave memory using `termsCommitment` as the key.

### BUY Handler

> `processWeatherBuy` decrypts ECIES ciphertext via the TEE node's `/decrypt` endpoint, validates `PrivateBuyParams`, stores terms in memory, and returns ABI-encoded parameters for `relayPrivateBuy`:

Reviewer: ```suggestion
Tje `processWeatherBuy` function decrypts ECIES ciphertext via the TEE node's `/decrypt` endpoint, validates `PrivateBuyParams`, stores terms in memory, and returns ABI-encoded parameters for `relayPrivateBuy`:
```

```go title="internal/extension/extension.go"
plaintext, err := decryptViaNode(e.signPort, df.OriginalMessage)
params, err := structs.Decode[types.PrivateBuyParams](types.PrivateBuyParamsArg, plaintext)
// ...
e.privateTerms[commitment] = params
encoded, err := types.PrivateBuyResultArgs.Pack(params.Holder, params.ContractAddr, ...)
return buildResult(action, df, encoded, 1, nil)
```

## Deploying and Testing on Coston2

> This walkthrough deploys a **local simulated TEE** against the real Coston2 chain using Docker and an ngrok tunnel.

Reviewer: ```suggestion
This walkthrough deploys a **local simulated TEE** against the real Coston2 chain using Docker and an `ngrok` tunnel.
```

For production deployment on a Google Cloud Platform Confidential Space VM, see `DEPLOYMENT_STEPS.md` in the repository.

### Step 0: Activate local simulated mode

```bash
./scripts/use-chain.sh local coston2
```

This copies `.env.local.coston2` to `.env`, setting `SIMULATED_TEE=true` and `LOCAL_MODE=false`.

### Step 1: Configure deployer keys and API key

Edit `.env.local.coston2` and set your funded Coston2 credentials and OpenWeatherMap key:

```bash title=".env.local.coston2"
DEPLOYMENT_PRIVATE_KEY="<your-funded-coston2-private-key-hex-no-0x>"
INITIAL_OWNER="0x<your-address>"
OPENWEATHERMAP_API_KEY="<your-openweathermap-api-key>"
PAY_TOKEN="0x53192e788991AD96bC180249B15AefB94E597dD1"
```

> `PAY_TOKEN` is a mocked [ERC-20](https://eips.ethereum.org/EIPS/eip-20) token on Coston2 — used for premiums and payouts during testing.

Reviewer: ```suggestion
The `PAY_TOKEN` variable is the address of a mocked [ERC-20](https://eips.ethereum.org/EIPS/eip-20) token on Coston2 — used for premiums and payouts during testing.
```

You can use this token for testing purposes or deploy your own ERC-20 token.

Re-activate after editing:

```bash
./scripts/use-chain.sh local coston2
```

### Step 2: Deploy contract and register extension

```bash
./scripts/pre-build.sh
./scripts/extension-setup.sh
```

> `pre-build.sh` compiles Solidity, deploys `WeatherInsurance`, and registers the extension on-chain.

Reviewer: ```suggestion
The `pre-build.sh` script compiles Solidity, deploys `WeatherInsurance`, and registers the extension on-chain.
```

It writes `EXTENSION_ID` and `INSTRUCTION_SENDER` to `config/extension.env`.

> `extension-setup.sh` calls `setPayToken` on the deployed contract.

Reviewer: ```suggestion
The `extension-setup.sh` script calls `setPayToken` on the deployed contract.
```

```

Re-activate:

```bash
./scripts/use-chain.sh local coston2
```

### Step 4: Configure the indexer database

```bash
cp config/proxy/extension_proxy.coston2.docker.toml.example \
   config/proxy/extension_proxy.coston2.docker.toml
```

Edit the `[db]` block with Coston2 indexer credentials — same process as the [Private Key Extension](/fcc/guides/sign-extension#step-4-configure-the-indexer-database).

:::info[Flare Indexer Access]
To get the indexer credentials, please get in touch with us via [support](https://flare.network/resources/technical-support) or [X](https://x.com/FlareDevs) and share what you are building.
:::

### Step 5: Start the extension stack

```bash
./scripts/start-services.sh
```

Wait for the proxy to become healthy:

```bash
until curl -sf http://localhost:6674/info >/dev/null 2>&1; do sleep 2; done
echo "Extension proxy is ready"
```

### Step 6: Verify the proxy

```bash
curl -s "$EXT_PROXY_URL/info" | jq '.machineData'
```

For a simulated TEE, expect the same `codeHash` and `extensionId` pattern as the sign extension.

### Step 7: Register the TEE machine

```bash
./scripts/post-build.sh
./scripts/extension-post-setup.sh
```

> `post-build.sh` allows the TEE version and registers the machine.

Reviewer: ```suggestion
The `post-build.sh` script allows the TEE version and registers the machine.
```

> `extension-post-setup.sh` reads the TEE signing address from the proxy and calls `setTeeAddress` on `WeatherInsurance`.

Reviewer: ```suggestion
The `extension-post-setup.sh` script reads the TEE signing address from the proxy and calls `setTeeAddress` on the `WeatherInsurance` contract.
```

### Step 8: Run the end-to-end test

```bash
./scripts/test.sh
```

> The test performs:

Reviewer: ```suggestion
The test does the following:
```
